File: scripts/python/spotify_client.py
"""
Spotify API Client

Provides a simple interface to the Spotify Web API with automatic
token management and error handling.
"""

import requests
from typing import Dict, Optional, Any
from spotify_auth import SpotifyAuth


class SpotifyClient:
    """
    Spotify API client with automatic token management.

    Handles authenticated requests to the Spotify Web API with
    automatic token refresh and comprehensive error handling.
    """

    BASE_URL = "https://api.spotify.com/v1"

    def __init__(self, auth: SpotifyAuth):
        """
        Initialize Spotify API client.

        Args:
            auth: SpotifyAuth instance for token management
        """
        self.auth = auth

    def _request(self, method: str, endpoint: str, params: Optional[Dict] = None) -> Any:
        """
        Make an authenticated request to the Spotify API.

        Args:
            method: HTTP method (GET, POST, etc.)
            endpoint: API endpoint (e.g., '/me/episodes')
            params: Query parameters

        Returns:
            Response JSON data

        Raises:
            Exception: On API errors with descriptive messages
        """
        # Get fresh token (auto-refreshes if needed)
        token = self.auth.get_access_token()

        headers = {
            'Authorization': f'Bearer {token}',
            'Content-Type': 'application/json'
        }

        url = f"{self.BASE_URL}{endpoint}"

        try:
            response = requests.request(
                method,
                url,
                headers=headers,
                params=params,
                timeout=10
            )

            # Handle HTTP errors
            if response.status_code == 401:
                raise Exception(
                    "Unauthorized: Invalid or expired Spotify access token. "
                    "Run 'python authenticate.py' to re-authenticate."
                )
            elif response.status_code == 403:
                raise Exception(
                    "Forbidden: Insufficient permissions. "
                    "Run 'python authenticate.py' to authenticate with correct scopes."
                )
            elif response.status_code == 404:
                error_msg = response.json().get('error', {}).get('message', 'Resource not found')
                raise Exception(f"Not found: {error_msg}")
            elif response.status_code == 429:
                raise Exception("Rate limited: Too many requests. Please wait before retrying.")
            elif response.status_code >= 500:
                raise Exception("Spotify API error: Server error. Please try again later.")

            response.raise_for_status()
            return response.json()

        except requests.exceptions.Timeout:
            raise Exception("Request timeout: Spotify API took too long to respond")
        except requests.exceptions.RequestException as e:
            if hasattr(e, 'response') and e.response is not None:
                error_data = e.response.json() if e.response.content else {}
                error_msg = error_data.get('error', {}).get('message', str(e))
                raise Exception(f"Spotify API error: {error_msg}")
            raise Exception(f"Request failed: {e}")

    def get_saved_episodes(
        self,
        limit: int = 20,
        offset: int = 0,
        market: Optional[str] = None
    ) -> Dict:
        """
        Get user's saved episodes.

        Args:
            limit: Maximum number of episodes to return (1-50)
            offset: Index of the first episode to return
            market: ISO 3166-1 alpha-2 country code (e.g., 'US')

        Returns:
            Paginated response with saved episodes

        API Endpoint: GET /me/episodes
        Required Scope: user-library-read
        """
        params = {
            'limit': limit,
            'offset': offset
        }

        if market:
            params['market'] = market

        return self._request('GET', '/me/episodes', params)

    def get_episode_details(
        self,
        episode_id: str,
        market: Optional[str] = None
    ) -> Dict:
        """
        Get detailed information about an episode.

        Args:
            episode_id: The Spotify ID for the episode
            market: ISO 3166-1 alpha-2 country code (e.g., 'US')

        Returns:
            Episode object with full metadata

        API Endpoint: GET /episodes/{id}
        """
        params = {}
        if market:
            params['market'] = market

        return self._request('GET', f'/episodes/{episode_id}', params if params else None)

    def get_show_details(self, show_id: str) -> Dict:
        """
        Get detailed information about a show/podcast.

        Args:
            show_id: The Spotify ID for the show

        Returns:
            Show object with full metadata

        API Endpoint: GET /shows/{id}
        """
        return self._request('GET', f'/shows/{show_id}')

    def search(
        self,
        query: str,
        types: str,
        limit: int = 20,
        offset: int = 0,
        market: Optional[str] = None
    ) -> Dict:
        """
        Search for items in the Spotify catalog.

        Args:
            query: Search query keywords
            types: Comma-separated list of item types (e.g., 'episode,show')
            limit: Maximum number of results (1-50)
            offset: Index of first result
            market: ISO 3166-1 alpha-2 country code (e.g., 'US')

        Returns:
            Search results

        API Endpoint: GET /search
        """
        params = {
            'q': query,
            'type': types,
            'limit': limit,
            'offset': offset
        }

        if market:
            params['market'] = market

        return self._request('GET', '/search', params)

File: scripts/server-src/services/spotifyClient.ts
import axios, { AxiosInstance, AxiosError } from "axios";
import {
  PaginatedResponse,
  SavedEpisodeObject,
  SpotifyEpisode,
  SpotifyShow,
} from "../types.js";
import { SpotifyAuth } from "../auth/spotifyAuth.js";

type TokenProvider = string | SpotifyAuth;

export class SpotifyClient {
  private client: AxiosInstance;
  private tokenProvider: TokenProvider;

  constructor(tokenProvider: TokenProvider) {
    this.tokenProvider = tokenProvider;

    this.client = axios.create({
      baseURL: "https://api.spotify.com/v1",
      headers: {
        "Content-Type": "application/json",
      },
      timeout: 10000,
    });

    // Add request interceptor to inject fresh token
    this.client.interceptors.request.use(async (config) => {
      const token = await this.getToken();
      config.headers.Authorization = `Bearer ${token}`;
      return config;
    });
  }

  /**
   * Get the current access token (refreshes if needed when using SpotifyAuth)
   */
  private async getToken(): Promise<string> {
    if (typeof this.tokenProvider === "string") {
      return this.tokenProvider;
    }
    return this.tokenProvider.getAccessToken();
  }

  async getSavedEpisodes(
    limit: number = 20,
    offset: number = 0,
    market?: string
  ): Promise<PaginatedResponse<SavedEpisodeObject>> {
    try {
      const params: Record<string, unknown> = {
        limit,
        offset,
      };

      if (market) {
        params.market = market;
      }

      const response = await this.client.get<
        PaginatedResponse<SavedEpisodeObject>
      >("/me/episodes", { params });

      return response.data;
    } catch (error) {
      throw this.handleError(error, "Failed to fetch saved episodes");
    }
  }

  async getEpisodeDetail(
    episodeId: string,
    market?: string
  ): Promise<SpotifyEpisode> {
    try {
      const params: Record<string, unknown> = {};

      if (market) {
        params.market = market;
      }

      const response = await this.client.get<SpotifyEpisode>(
        `/episodes/${episodeId}`,
        { params }
      );

      return response.data;
    } catch (error) {
      throw this.handleError(error, `Failed to fetch episode ${episodeId}`);
    }
  }

  async getShowDetails(showId: string): Promise<SpotifyShow> {
    try {
      const response = await this.client.get<SpotifyShow>(
        `/shows/${showId}`
      );

      return response.data;
    } catch (error) {
      throw this.handleError(error, `Failed to fetch show ${showId}`);
    }
  }

  async search(
    query: string,
    types: string,
    limit: number = 20,
    offset: number = 0,
    market?: string
  ): Promise<unknown> {
    try {
      const params: Record<string, unknown> = {
        q: query,
        type: types,
        limit,
        offset,
      };

      if (market) {
        params.market = market;
      }

      const response = await this.client.get<unknown>(
        "/search",
        { params }
      );

      return response.data;
    } catch (error) {
      throw this.handleError(error, "Failed to search catalog");
    }
  }

  private handleError(
    error: unknown,
    defaultMessage: string
  ): never {
    if (axios.isAxiosError(error)) {
      const axiosError = error as AxiosError<{ error?: { message?: string } }>;

      if (axiosError.response?.status === 401) {
        throw new Error(
          "Unauthorized: Invalid or expired Spotify access token. Run 'npm run auth' to re-authenticate."
        );
      }

      if (axiosError.response?.status === 403) {
        throw new Error(
          "Forbidden: Insufficient permissions. Run 'npm run auth' to authenticate with correct scopes."
        );
      }

      if (axiosError.response?.status === 404) {
        throw new Error(`Not found: ${axiosError.response.data?.error?.message || "Resource not found"}`);
      }

      if (axiosError.response?.status === 429) {
        throw new Error(
          "Rate limited: Too many requests. Please wait before retrying."
        );
      }

      if (axiosError.response?.status === 500) {
        throw new Error(
          "Spotify API error: Server error. Please try again later."
        );
      }

      if (axiosError.message === "timeout of 10000ms exceeded") {
        throw new Error("Request timeout: Spotify API took too long to respond");
      }

      throw new Error(
        `${defaultMessage}: ${axiosError.response?.data?.error?.message || axiosError.message}`
      );
    }

    throw new Error(`${defaultMessage}: ${String(error)}`);
  }
}
